internal/jobcompleter/job_completer.go

```diff
@@ -56,125 +63,299 @@ type InlineJobCompleter struct {
+
+		c.Logger.InfoContext(ctx, c.Name+": Run loop started")
+		defer c.Logger.InfoContext(ctx, c.Name+": Run loop stopped")
 
-		func() {
-			c.subscribeFuncMu.Lock()
-			defer c.subscribeFuncMu.Unlock()
+		c.debounceChan = chanutil.NewDebouncedChan(stopCtx, 100*time.Millisecond)
 
-			if c.subscribeFunc != nil {
-				c.subscribeFunc(CompleterJobUpdated{Job: job, JobStats: stats})
+		close(c.ready)
+
+		for {
+			select {
+			case <-stopCtx.Done():
+				// Try to insert last batch before leaving. Note we use the
+				// original context so operations aren't immediately cancelled.
+				if err := c.handleBatch(ctx); err != nil {
+					c.Logger.Error(c.Name+": Error completing batch", "err", err)
+				}
+
+				return
+			case <-c.debounceChan.C():
+				if err := c.handleBatch(ctx); err != nil {
+					c.Logger.Error(c.Name+": Error completing batch", "err", err)
+				}
 			}
-		}()
+		}
+	}()
+
+	return nil
+}
 
+func (c *BatchCompleter) handleBatch(ctx context.Context) error {
+	var setStateBatch map[int64]*batchCompleterSetState
+	func() {
+		c.setStateParamsMu.Lock()
+		defer c.setStateParamsMu.Unlock()
+
+		setStateBatch = c.setStateParams
+
+		// Don't bother resetting the map if there's nothing to process,
+		// allowing the completer to idle efficiently.
+		if len(setStateBatch) > 0 {
+			c.setStateParams = make(map[int64]*batchCompleterSetState)
+		} else {
+			// Set nil to avoid a data race below in case the map is set as a
+			// new job comes in.
+			setStateBatch = nil
+		}
+	}()
+
+	if len(setStateBatch) < 1 {
 		return nil
-	})
+	}
+
+	// Insert a sub-batch with retries. Also helps reduce visual noise and
+	// increase readability of loop below.
+	insertSubBatch := func(batchID []int64, batchFinalizedAt []time.Time) ([]*rivertype.JobRow, error) {
+		return withRetries(ctx, &c.BaseService, func(ctx context.Context) ([]*rivertype.JobRow, error) {
+			return c.exec.JobSetCompleteIfRunningMany(ctx, &riverdriver.JobSetCompleteIfRunningManyParams{
+				ID:          batchID,
+				FinalizedAt: batchFinalizedAt,
+			})
+		})
+	}
+
+	// This could be written more simply using multiple `sliceutil.Map`s, but
+	// it's done this way to allocate as few new slices as necessary.
+	mapIDsAndFinalizedAt := func(setStateBatch map[int64]*batchCompleterSetState) ([]int64, []time.Time) {
+		var (
+			batchIDs         = make([]int64, len(setStateBatch))
+			batchFinalizedAt = make([]time.Time, len(setStateBatch))
+			i                int
+		)
+		for _, setState := range setStateBatch {
+			batchIDs[i] = setState.Params.ID
+			batchFinalizedAt[i] = *setState.Params.FinalizedAt
+			i++
+		}
+		return batchIDs, batchFinalizedAt
+	}
+
+	// Tease apart enormous batches into sub-batches.
+	//
+	// All the code below is concerned with doing that, with a fast loop that
+	// doesn't allocate any additional memory in case the entire batch is
+	// smaller than the sub-batch maximum size (which will be the common case).
+	const oneOperationMax = 2_000
```

Comment:
Maybe you want to make this a configurable field on the completer type? I guess we don't have any batch completer specific testing atm, but it might make such tests easier to write.

Reply:
I ended up extracting this constant and putting it on the struct for testability, and added a few specific tests for the batch completer to exercise these edges.